### PR TITLE
init commit

### PR DESCRIPTION
**Pull request by calvix**
initial commit of separating service resources from operators
related to https://github.com/giantswarm/giantswarm/issues/3248

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,2 @@
 # serviceresource
- Package serviceresource implements primitives for service used for monitoring within Giant Swarm Kubernetes guest clusters. 
+ Package serviceresource implements primitives for service used within Giant Swarm Kubernetes clusters.
```
Comment (xh3b4sd) [inline README.md:2]: The readme badge is missing.

File: current.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/finalizerskeptcontext"
+	"github.com/giantswarm/operatorkit/controller/context/resourcecanceledcontext"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error) {
+	clusterID, err := clusterID(obj)
```
Comment (xh3b4sd) [inline current.go:16]: This has to be configurable so the provider operators can set their own key functions.

File: current.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+	var services []*corev1.Service
+
+	namespace := clusterNamespace(clusterID)
+	serviceNames := []string{
+		serviceNameMaster,
+	}
+	if isKVM(obj) {
+		serviceNames = append(serviceNames, serviceNameWorker)
+	}
```
Comment (xh3b4sd) [inline current.go:31]: This has to be configurable so the provider operators can set their own key functions.

File: current.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d services in the Kubernetes API", len(services)))
+
+	// In case a cluster deletion happens, we want to delete the guest cluster
+	// services. We still need to use the services for ingress routing in order to
+	// drain nodes on KVM though. So as long as pods are there we delay the
+	// deletion of the services here in order to still be able to route traffic to
+	// the guest cluster API. As soon as the draining was done and the pods got
+	// removed we get an empty list here after the delete event got replayed. Then
+	// we just remove the services as usual.
+	if isKVM(obj) {
```
Comment (xh3b4sd) [inline current.go:55]: This is an abstracted resource. Rule of thumb is, grep for the provider in this repo and when you find any result something is wrong. The abstraction we need here apparently is to have configurable hooks.

File: desired.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package service
+
+import (
+	"context"
+)
+
+func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
+	return r.desiredServicesFunc(ctx, obj)
```
Comment (xh3b4sd) [inline desired.go:8]: Please lets keep the calls explicit for our error masking. The masking tells a story via its stack. This story has to match the real call graph and your mental model. This here breaks it.

File: resource.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// Name is the identifier of the resource.
+	Name = "servicev14"
```
Comment (xh3b4sd) [inline resource.go:15]: This should probably just be `serviceresource`.

File: resource.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+type Config struct {
+	DesiredServicesFunc func(ctx context.Context, obj interface{}) (*[]apiv1.Service, error)
+	// Dependencies.
+	K8sClient kubernetes.Interface
+	Logger    micrologger.Logger
+}
+
+// DefaultConfig provides a default configuration to create a new service
+// resource by best effort.
+func DefaultConfig() Config {
```
Comment (xh3b4sd) [inline resource.go:28]: We do not maintain these anymore.

File: resource.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	return Config{
+		// Dependencies.
+		K8sClient: nil,
+		Logger:    nil,
+	}
+}
+
+// Resource implements the service resource.
+type Resource struct {
+	desiredServicesFunc func(ctx context.Context, obj interface{}) (*[]apiv1.Service, error)
```
Comment (xh3b4sd) [inline resource.go:38]: This is a dependency because it implements behaviour.

**Review by xh3b4sd — CHANGES_REQUESTED** [review]
This PR is too big and does too many things at once. Ideally we would deal with these separate PRs. 

- Circle setup. 
- Resource structure without functionality. 
- Functionality of GetCurrentState. 
- Functionality of GetDesiredState.
- ....

**Comment by calvix** [comment]
I don't understand half of the comments, I don't think its still a good idea that I should continue here.

**Comment by calvix** [comment]
closing for now, but not deleting - https://github.com/giantswarm/giantswarm/issues/3748

**Pull request closed without merging**